RecoTracker/PixelSeeding/plugins/alpaka/CAGeometryESProucer.cc

```diff
@@ -0,0 +1,207 @@
+    subSystemName = GeomDetEnumerators::tkDetEnum[++subSystem];
+    subSystemOffset = trackerGeometry->offsetDU(subSystemName);
+#endif
+
+    auto oldLayer = 0u;
+    auto layerCount = 0;
+
+    std::vector<int> layerStarts(n_layers + 1);
+
+    for (auto& det : dets) {
+      DetId detid = det->geographicalId();
+#ifdef GPU_DEBUG
+      if (n_modules >= int(subSystemOffset)) {
+        subSystemName = GeomDetEnumerators::tkDetEnum[++subSystem];
+        subSystemOffset = trackerGeometry->offsetDU(subSystemName);
+        std::cout << " ===================== Subsystem: " << subSystemName << std::endl;
+      }
+#endif
+
+      auto layer = trackerTopology->layer(detid);
+
+      if (layer != oldLayer) {
+        layerStarts[layerCount++] = n_modules;
+
+        if (layerCount > n_layers + 1)
+          break;
+
+        oldLayer = layer;
+#ifdef GPU_DEBUG
+        std::cout << " > New layer at module : " << n_modules << " (detId: " << detid << ")" << std::endl;
+#endif
+      }
+
+      n_modules++;
+    }
+
+    reco::CAGeometryHost product{{{n_layers + 1, n_pairs, n_modules}}, cms::alpakatools::host()};
+
+    auto layerSoA = product.view();
+    auto cellSoA = product.view<::reco::CAGraphSoA>();
+    auto modulesSoA = product.view<::reco::CAModulesSoA>();
+
+    for (int i = 0; i < n_modules; ++i) {
+      auto det = dets[i];
+      auto vv = det->surface().position();
+      auto rr = Rotation(det->surface().rotation());
+      modulesSoA[i].detFrame() = Frame(vv.x(), vv.y(), vv.z(), rr);
+    }
+
+    for (int i = 0; i < n_layers; ++i) {
+      layerSoA.layerStarts()[i] = layerStarts[i];
+      layerSoA.caThetaCut()[i] = caThetaCuts_[i];
+      layerSoA.caDCACut()[i] = caDCACuts_[i];
+    }
+
+    layerSoA.layerStarts()[n_layers] = layerStarts[n_layers];
+
+    for (int i = 0; i < n_pairs; ++i) {
+      cellSoA.graph()[i] = {{uint32_t(pairGraph_[2 * i]), uint32_t(pairGraph_[2 * i + 1])}};
+      cellSoA.phiCuts()[i] = phiCuts_[i];
+      cellSoA.minz()[i] = minZ_[i];
+      cellSoA.maxz()[i] = maxZ_[i];
+      cellSoA.maxr()[i] = maxR_[i];
+      cellSoA.startingPair()[i] = false;
+    }
+
+    for (const int& i : startingPairs_)
+      cellSoA.startingPair()[i] = true;
+
+    return product;
+  }
+
+  void CAGeometryESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    using namespace phase1PixelTopology;
+    constexpr auto nPairs = pixelTopology::Phase1::nPairsForQuadruplets;
+    edm::ParameterSetDescription desc;
+
+    // layers params
+    desc.add<std::vector<double>>("caDCACuts", {0.15f, 0.25f, 0.25f, 0.25f, 0.25f, 0.25f, 0.25f, 0.25f, 0.25f, 0.25f})
+        ->setComment("Cut on RZ alignement. One per layer, the layer being the midelle one for a triplet.");
+    desc.add<std::vector<double>>("caThetaCuts",
+                                  {0.002f, 0.002f, 0.002f, 0.002f, 0.003f, 0.003f, 0.003f, 0.003f, 0.003f, 0.003f})
```

Comment:
These are vector<double> but you use float numbers to initialise - why?

Reply:
what's wrong here, sorry?
`{0.002f, 0.002f, 0.002f, 0.002f, 0.003f, 0.003f, 0.003f, 0.003f, 0.003f, 0.003f}`  **is** a vector, see e.g. [here](https://alexsm.com/cpp-std-vector-braced-init/).

Reply:
What I mean is that it is a vector of float, not a vector of double. So there is conversion from float (constants) elements to double elements to make a vector of double returned by the getParameter call.
But I see this pattern is already used elsewhere, I wonder why...

Reply:
If I am not mistaken, the framework doesn't like `float` parameters, they must be `double`. But then these inputs will be actually stored in a `float`

```c++
 GENERATE_SOA_LAYOUT(CALayersLayout,
                      SOA_COLUMN(uint32_t, layerStarts),
                      SOA_COLUMN(float, caThetaCut),
                      SOA_COLUMN(float, caDCACut))
```

So the initialization with `float`s is just to avoid (hopefully) funny decimal tails when casting them from double to floats. I don't think it has much of an influence anyway. For uniformity, I should expand this also to the other parameters below actually.

Reply:
Felice opened https://github.com/cms-sw/cmssw/pull/45216 to add support for floats, but it's been sleeping for a while.